File: wpimath/src/main/native/cpp/system/plant/struct/DCMotorStruct.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "wpi/math/system/plant/struct/DCMotorStruct.hpp"

namespace {
constexpr size_t kNominalVoltageOff = 0;
constexpr size_t kStallTorqueOff = kNominalVoltageOff + 8;
constexpr size_t kStallCurrentOff = kStallTorqueOff + 8;
constexpr size_t kFreeCurrentOff = kStallCurrentOff + 8;
constexpr size_t kFreeSpeedOff = kFreeCurrentOff + 8;
}  // namespace

using StructType = wpi::Struct<frc::DCMotor>;

frc::DCMotor StructType::Unpack(std::span<const uint8_t> data) {
  return frc::DCMotor{
      units::volt_t{wpi::UnpackStruct<double, kNominalVoltageOff>(data)},
      units::newton_meter_t{wpi::UnpackStruct<double, kStallTorqueOff>(data)},
      units::ampere_t{wpi::UnpackStruct<double, kStallCurrentOff>(data)},
      units::ampere_t{wpi::UnpackStruct<double, kFreeCurrentOff>(data)},
      units::radians_per_second_t{
          wpi::UnpackStruct<double, kFreeSpeedOff>(data)},
  };
}

void StructType::Pack(std::span<uint8_t> data, const frc::DCMotor& value) {
  wpi::PackStruct<kNominalVoltageOff>(data, value.nominalVoltage.value());
  wpi::PackStruct<kStallTorqueOff>(data, value.stallTorque.value());
  wpi::PackStruct<kStallCurrentOff>(data, value.stallCurrent.value());
  wpi::PackStruct<kFreeCurrentOff>(data, value.freeCurrent.value());
  wpi::PackStruct<kFreeSpeedOff>(data, value.freeSpeed.value());
}

File: wpimath/src/main/java/edu/wpi/first/math/system/plant/struct/DCMotorStruct.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package edu.wpi.first.math.system.plant.struct;

import edu.wpi.first.math.system.plant.DCMotor;
import edu.wpi.first.util.struct.Struct;
import java.nio.ByteBuffer;

public class DCMotorStruct implements Struct<DCMotor> {
  @Override
  public Class<DCMotor> getTypeClass() {
    return DCMotor.class;
  }

  @Override
  public String getTypeName() {
    return "DCMotor";
  }

  @Override
  public int getSize() {
    return kSizeDouble * 5;
  }

  @Override
  public String getSchema() {
    return "double nominal_voltage;double stall_torque;double stall_current;"
        + "double free_current;double free_speed";
  }

  @Override
  public DCMotor unpack(ByteBuffer bb) {
    double nominalVoltage = bb.getDouble();
    double stallTorque = bb.getDouble();
    double stallCurrent = bb.getDouble();
    double freeCurrent = bb.getDouble();
    double freeSpeed = bb.getDouble();
    return new DCMotor(nominalVoltage, stallTorque, stallCurrent, freeCurrent, freeSpeed, 1);
  }

  @Override
  public void pack(ByteBuffer bb, DCMotor value) {
    bb.putDouble(value.nominalVoltageVolts);
    bb.putDouble(value.stallTorqueNewtonMeters);
    bb.putDouble(value.stallCurrentAmps);
    bb.putDouble(value.freeCurrentAmps);
    bb.putDouble(value.freeSpeedRadPerSec);
  }
}
